### ISSUE TITLE
attention performance regression from switch to torch 2.5.0/flex attention on MI-300

### ISSUE THREAD

**Issue by jonabur** [pytorch/torchtune#1879]
I am testing performance of training with a packed dataset, and upgrading to torch 2.5.0 causes about a 30% performance tokens/gpu/second drop that looks like it is related to the switch to flex attention, especially the triton kernel during the backwards pass.  Prior to the torch upgrade packed datasetws were using the sdpa math backend.

Here is an example command that has a performance difference between 2.5 and 2.4 (optimizer.fused=False because it was causing an error on multi-device training on torch 2.4 for me)

```bash
tune run --nproc_per_node 8 \
    lora_finetune_distributed \
    --config llama3_1/8B_lora \
    batch_size=5 \
    gradient_accumulation_steps=1 \
    tokenizer.max_seq_len=4096 \
    optimizer.fused=False \
    dataset.packed=True
```
Yields:
Torch 2.4: ~5500 t/g/s
Torch 2.5: ~3700 t/g/s

A trace showed the triton kernel launched from CompiledFunctionBackward taking >100ms, which looks to be substantially more time spent per layer backward pass than on the 2.4 sdpa math backend (it looks like maybe 3x longer?)

I tried to disable Flexattention to verify that the performance change is specific to FlexAttention. There's no option for this, so I tried to disable it by hardcoding the _SUPPORTS_FLEX_ATTENTION to False in torchtune/utils/_import_guard.py, but this caused NaN loss during training. I have not had time to investigate this approach further yet.

Is this kind of performance difference expected with flex attention?
In either case, is it possible to have an option to disable flex attention, at least for now?

**Comment by jonabur** [comment]
Nevermind, I think i have an environment issue here.  Will do more investigation and reopen if necessary.
